llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -905,6 +905,24 @@ AArch64TTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
     }
     break;
   }
+  case Intrinsic::experimental_vector_match: {
+    if (auto *NeedleTy = dyn_cast<FixedVectorType>(ICA.getArgTypes()[1])) {
```

Comment:
Any reason this cannot be just `auto *NeedleTy = cast<FixedVectorType>(ICA.getArgTypes()[1])` without needing the extra control flow?

Reply:
I believe `-type-based-intrinsic-cost` calls into this with a non-vector argument.

Reply:
I would expect that only for overloaded intrinsics that support both scalar and vector operations. Are you seeing a real failure because of a bogus type?

`vector_reverse` is another example of a vector only intrinsic where I can see the base implementation of `getIntrinsicInstrCost` is happy to assume the operand type is what it expects it to be. I'd kind of hope the cost model is allow to assume the IR is wellformed.

Reply:
Without the type check that line was running into a segfault when called from [here](https://github.com/llvm/llvm-project/blob/53e9eee0e2c9587fc7a1e7b5f64ba2a935f58e0a/llvm/include/llvm/CodeGen/BasicTTIImpl.h#L2024-L2026). It wasn't obvious to me I was missing anything elsewhere, so I thought guarding the case was the correct way of approaching this. Did I miss anything elsewhere?

Reply:
Thanks very much for the suggestion, I've attempted that in 15df9dd2428cc498fe59f4ff8d9a9e3c9200d8d5. Could you please have a look if that looks better?